### PR TITLE
feat: separate out VolOracleLibrary, cover more edge cases for binarySearch, add test

### PR DESCRIPTION
**Pull request by PangZhi**

### CODE REVIEW

File: contracts/VolOracle.sol [hunk 1]
```diff
@@ -2,45 +2,23 @@
...
-    struct VolObservation {
-        // the block timestamp of the observation
-        uint32 blockTimestamp;
-        // the tick accumulator, i.e. tick * time elapsed since the pool was first initialized
-        int56 tickCumulative;
-        // the tick square accumulator, i.e. tick * tick * time elapsed since the oracle was first initialized
-        uint112 tickSquareCumulative;
-    }
+    // usiVolOracleLibary for VolOracleState;
+    using VolOracleLib for VolOracleLib.VolObservation;
```
Comment (ankitchiplunkar) [inline contracts/VolOracle.sol]: is this needed?

File: contracts/VolOracleLib.sol [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        uint256 startIndex = getObservationIndexBeforeOrAtTarget(self, target);
+        uint256 endIndex = self.observationIndex;
+        require(startIndex != endIndex, "no new observations to calculate the volatility");
+        VolOracleLib.VolObservation memory startObservation = self.observations[startIndex];
+        VolOracleLib.VolObservation memory endObservation = self.observations[endIndex];
+        uint256 tickSquareSum = endObservation.tickSquareCumulative - startObservation.tickSquareCumulative;
+        uint256 timeEclapsed = uint256(endObservation.blockTimestamp - startObservation.blockTimestamp);
+        uint256 tickAvg = uint256(uint56(endObservation.tickCumulative - startObservation.tickCumulative)) /
+            timeEclapsed;
+        uint256 stddev = (tickSquareSum - tickAvg * tickAvg * timeEclapsed) / (timeEclapsed - 1);
```
Comment (ankitchiplunkar) [inline contracts/VolOracleLib.sol]:
```suggestion
       return (tickSquareSum - tickAvg * tickAvg * timeEclapsed) / (timeEclapsed - 1);
```

File: contracts/VolOracle.sol (hunk 1 above)
Comment (PangZhi) [reply contracts/VolOracle.sol]: ah probably only need for oracleState

File: contracts/VolOracleLib.sol [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        return stddev;
+    }
+
+    // @dev get the obesrvation index right before or at the target timestamp, using binary search
+    // @return the index of the observation which is before or at the target timestamp
+    function getObservationIndexBeforeOrAtTarget(VolOracleState storage self, uint32 _target)
```
Comment (ankitchiplunkar) [inline contracts/VolOracleLib.sol:44]: we can turn this into `internal` would be better for testing
